bgp/routes.go

```diff
@@ -0,0 +1,287 @@
+
+package bgp
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/netip"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"text/tabwriter"
+	"time"
+
+	"github.com/osrg/gobgp/v3/pkg/packet/bgp"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/bgpv1/api"
+	"github.com/cilium/cilium/pkg/bgpv1/types"
+
+	"github.com/cilium/cilium-cli/defaults"
+	"github.com/cilium/cilium-cli/status"
+)
+
+const (
+	availableKW  = "available"
+	advertisedKW = "advertised"
+	vRouterKW    = "vrouter"
+	peerKW       = "peer"
+	neighborKW   = "neighbor"
+	ipv4AFI      = "ipv4"
+	unicastSAFI  = "unicast"
+)
+
+// GetRoutes gets BGP routes from all/specific cilium agent pods.
+func (s *Status) GetRoutes(ctx context.Context, args []string) error {
+	silent := s.params.Output == status.OutputJSON // do not print out notes / warnings when the output is JSON
+	if len(args) < 1 {
+		args = defaultGetRoutesArgs(silent)
+	}
+	err := validateGetRoutesArgs(args)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancelFn := context.WithTimeout(ctx, s.params.WaitDuration)
+	defer cancelFn()
+
+	err = s.initTargetCiliumPods(ctx)
+	if err != nil {
+		return err
+	}
+
+	res, err := s.fetchRoutesConcurrently(ctx, args)
+	if err != nil {
+		if len(res) == 0 {
+			// no results retrieved - just return the error
+			return err
+		}
+		// print the errors, but continue with printing results
+		fmt.Fprintf(os.Stderr, "Errors by retrieving routes: %v\n\n", err)
+	}
+	return s.writeRoutes(res)
+}
+
+func (s *Status) fetchRoutesConcurrently(ctx context.Context, args []string) (map[string][]*models.BgpRoute, error) {
+	allFetchedData := make(map[string][]*models.BgpRoute)
+
+	// res contains data returned from cilium pod
+	type res struct {
+		nodeName string
+		data     []*models.BgpRoute
+		err      error
+	}
+	resCh := make(chan res)
+
+	var wg sync.WaitGroup
+
+	// max number of concurrent go routines will be number of cilium agent pods
+	wg.Add(len(s.ciliumPods))
+
+	// concurrently fetch routes from each cilium pod
+	for _, pod := range s.ciliumPods {
+		go func(ctx context.Context, pod *corev1.Pod) {
+			defer wg.Done()
+
+			routes, err := s.fetchRoutesFromPod(ctx, args, pod)
+			resCh <- res{
+				nodeName: pod.Spec.NodeName,
+				data:     routes,
+				err:      err,
+			}
+		}(ctx, pod)
+	}
+
+	// close resCh when data from all nodes is collected
+	go func() {
+		wg.Wait()
+		close(resCh)
+	}()
+
+	// read from the channel till it is closed.
+	// on error, store error and continue to next node.
+	var err error
+	for fetchedData := range resCh {
+		if fetchedData.err != nil {
+			err = errors.Join(err, fetchedData.err)
+		} else {
+			allFetchedData[fetchedData.nodeName] = fetchedData.data
+		}
+	}
+
+	return allFetchedData, err
+}
+
+func (s *Status) fetchRoutesFromPod(ctx context.Context, args []string, pod *corev1.Pod) ([]*models.BgpRoute, error) {
+	cmd := []string{"cilium-dbg", "bgp", "routes"}
```

Comment:
Do we need to recompute the static cmd for every pods?

Reply:
good point, fixed